File: output/ChatGPT/avatar/Java/C++/atcoder_AGC046_B.cpp
#include <iostream>
#include <vector>

using namespace std;

const int MOD = 998244353;

void solve() {
    int A, B, C, D;
    cin >> A >> B >> C >> D;

    vector<vector<long long>> dp(C + 1, vector<long long>(D + 1, 0));
    dp[A][B] = 1;

    for (int c = A; c <= C; c++) {
        for (int d = B; d <= D; d++) {
            if (c == A && d == B) {
                continue;
            }
            long long ans = 0;

            if (c > A) {
                long long part = dp[c - 1][d];
                ans = (part * d) % MOD;
            }
            if (d > B) {
                long long part = dp[c][d - 1];
                ans = (ans + (part * c) % MOD) % MOD;
            }
            if (c > A && d > B) {
                ans = (ans - (dp[c - 1][d - 1] * (c - 1) % MOD * (d - 1) % MOD) + MOD) % MOD;
            }

            dp[c][d] = ans;
        }
    }

    cout << dp[C][D] << endl;
}

int main() {
    solve();
    return 0;
}

File: output/ChatGPT/avatar/Java/Python/atcoder_AGC046_B.py
# Converted Java code to Python

MOD = 998244353

dp = [[0] * 3001 for _ in range(3001)]

def solve(A, B, C, D):
    dp[A][B] = 1
    for c in range(A, C+1):
        for d in range(B, D+1):
            if c == A and d == B:
                continue
            ans = 0
            if c > A:
                part = dp[c-1][d]
                ans = (part * d) % MOD
            if d > B:
                part = dp[c][d-1]
                ans = (ans + (part * c)) % MOD
            if c > A and d > B:
                ans = (ans - (dp[c-1][d-1] * (c-1) * (d-1))) % MOD
            dp[c][d] = (ans + MOD) % MOD

    return dp[C][D]

A, B, C, D = map(int, input().split())
result = solve(A, B, C, D)
print(result)